docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,298 @@
+---
+title: "Exported TypeScript Clients Across Wing Applications"
+description: Exported TypeScript Clients Across Wing Applications
+---
+
+# Exported TypeScript Clients Across Wing Applications
+
+- **Author(s)**: @MarkMcCulloh
+- **Submission Date**: 2024-06-02
+
+## Use Case
+
+Given some infrastructure defined in wing by platform engineers, we want to provide a way for application developers in the same organization to interact with that infrastructure in a type-safe way from their own codebases without dealing with wing or infrastructure themselves.
+
+As a platform engineer I'd like to use wing to create the preflight infrastructure for developers in my organization, and then produce client libraries in TypeScript for accessing that infrastructure within other pieces of code.
+Imagine for example that the application developers write TypeScript running in containers in their own repo, and that repo's code is already being automatically deployed with GitOps via ArgoCD or other means.
+
+As an application developer I'd like to consume the TypeScript package published by the platform team and write application code that interacts with these resources (using whichever methods or permissions have been granted by the platform team). If I'm running my TypeScript code / container locally I expect the wing clients to connect to the wing simulator and interact with the simulated resources, and when my code is deployed to the AWS I expect the clients to connect to the real cloud resources.
+
+## Proposal
+
+The `clients` module has a `TypeScript` class used to create a client. When this is instantiated, `wing compile` will produce a publishable JavaScript npm package with TypeScript type declarations.
+
+```wing
+bring clients;
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  // version: "0.0.0",
+  // outdir: join(@dirname, "dist"),
+);
+```
+
+To lift a resource into the client to make it accessible to consumers, add a `lift` expression:
+
+```wing
+bring cloud;
+
+let bucket = new cloud.Bucket();
+
+client.lift(lift { bucket: [put] });
+```
+
+In order to access the bucket, the client needs to know it should resolve the relevent information (e.g. bucket name or handle).
+This information is stored in client.context, so we need to way to both store and resolve this information.
+For example, we can store/retrieve the context from a file:
+
+```wing
+bring cloud;
+bring clients;
+bring fs;
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return fs.readJson("/var/run/wing-client-context.json");
+  },
```

Comment:
Not sure if there should be a default implementation, I feel like there are too many options to pick one

Reply:
Not sure where this inflight closure will be executed. 

I think we should simply make `context` a string token and then it can be referenced.